pkg/riskanalysis/cmd.go

```diff
@@ -92,80 +95,216 @@ func (opt *Options) Run() error {
+		return nil
+	}
+
+	return nil
+}
+
+// struct that records the timing and status of each RA http client request
+type raRequestLog struct {
+	RequestCount int // which iteration are we on for this job requesting RA
+	StartTime    time.Time
+	Duration     time.Duration
+	Error        string
+	BytesRead    int
+}
+
+// readWriteRiskAnalysis requests Risk Analysis from sippy, writes the results to disk, and returns the RA html to include in prow job output.
+// If the request fails, it will try up to maxTries times before returning an error; an error means no RA data returned.
+func (opt *Options) readWriteRiskAnalysis(inputBytes []byte) ([]byte, error) {
 	req, err := http.NewRequest("GET", opt.SippyURL, bytes.NewBuffer(inputBytes))
 	if err != nil {
 		logrus.WithError(err).Error("Error creating GET request during risk analysis")
-		return nil
+		return nil, err
 	}
 	req.Header.Set("Content-Type", "application/json")
-	client := &http.Client{}
+	riskAnalysisBytes, err := opt.requestRiskAnalysis(req, &http.Client{}, &realSleeper{})
+	if err != nil {
+		return nil, err
+	}
 
+	outputFile := filepath.Join(opt.JUnitDir, raDataFile)
+	err = os.WriteFile(outputFile, riskAnalysisBytes, 0644)
+	if err != nil {
+		logrus.WithError(err).Error("Error writing risk analysis json artifact")
+	} else {
+		logrus.Infof("Successfully wrote: %s", outputFile)
+	}
+
+	opt.writeRATestResults(riskAnalysisBytes)
+	return riskAnalysisBytes, nil // whether or not the file was written
+}
+
+// sleeper interface to enable testing without actually sleeping
+type sleeper interface {
+	Sleep(d time.Duration)
+}
+type realSleeper struct{}
+
+func (rs *realSleeper) Sleep(d time.Duration) {
+	time.Sleep(d)
+}
+
+// requestRiskAnalysis makes the http request(s) and records the timing and status for each
+func (opt *Options) requestRiskAnalysis(req *http.Request, client *http.Client, sleepy sleeper) ([]byte, error) {
 	var resp *http.Response
+	var err error
+	reqLogs := []*raRequestLog{}
+	var finalReqLog *raRequestLog = nil    // keep final log entry to amend before writing if needed
+	defer opt.writeRARequestLogs(&reqLogs) // write all failures or successes after processing
 	clientDoSuccess := false
-	for i := 1; i <= maxRetries; i++ {
+	for i := 1; i <= maxTries; i++ {
+		reqLog := &raRequestLog{RequestCount: i, StartTime: time.Now()}
+		finalReqLog = reqLog
+		reqLogs = append(reqLogs, finalReqLog)
 		ctx, cancelFn := context.WithTimeout(req.Context(), 20*time.Second)
 		defer cancelFn()
-		startTime := time.Now()
-		logrus.Infof("Requesting risk analysis (attempt %d/%d) from: %s", i, maxRetries, sippyURL)
+
+		logrus.Infof("Requesting risk analysis (attempt %d/%d) from: %s", i, maxTries, sippyURL)
 		resp, err = client.Do(req.WithContext(ctx))
-		endTime := time.Now()
-		duration := endTime.Sub(startTime)
-		logrus.Infof("Call to sippy finished after: %s", duration)
+		cancelFn() // cancel the context timeout we just used.
```

Comment:
Looks like you call this via defer on line 189, is it intentional to call it again?

Reply:
Ok, looks like the same call was previously at line 121.

Reply:
it's how it was when i started, just moved slightly. i wasn't sure i understood the context well enough to change. i gather it's to clean up the timeout, and if it hasn't already timed out, keep it from firing in the middle of the rest of the code (but what would happen if it did?). the defer does seem superfluous as there's no way it could make it through an iteration without calling `cancelFn()` but... calling it twice seems unlikely to cause problems?